ClosedXML/Excel/CalcEngine/Functions/Lookup.cs

```diff
@@ -87,6 +88,84 @@ private static object Hyperlink(List<Expression> p)
             return new XLHyperlink(address, toolTip);
         }
 
+        private static object Match(List<Expression> p)
+        {
+            var lookup_value = p[0];
+            int match_type = 1;
+            if (p.Count > 2)
+                match_type = Math.Sign((int)p[2]);
+
+            if (!(p[1] is XObjectExpression objectExpression))
+                throw new NoValueAvailableException("lookup_array has to be a range");
+
+            if (!(objectExpression.Value is CellRangeReference cellRangeReference))
+                throw new NoValueAvailableException("lookup_array has to be a range");
+
+            var range = cellRangeReference.Range;
+
+            if (range.ColumnCount() != 1 && range.RowCount() != 1)
+                throw new CellValueException("Range has to be 1-dimensional");
+
+            Predicate<int> lookupPredicate = null;
+            switch (match_type)
+            {
+                case 0:
+                    lookupPredicate = i => i == 0;
+                    break;
+
+                case 1:
+                    lookupPredicate = i => i <= 0;
+                    break;
+
+                case -1:
+                    lookupPredicate = i => i >= 0;
+                    break;
+
+                default:
+                    throw new NoValueAvailableException("Invalid match_type");
+            }
+
+            IXLCell foundCell = null;
+
+            if (match_type == 0)
+                foundCell = range
+                    .CellsUsed(XLCellsUsedOptions.Contents, c => !c.IsEmpty() && lookupPredicate.Invoke(new Expression(c.Value).CompareTo(lookup_value)))
```

Comment:
` .CellsUsed(XLCellsUsedOptions.Contents, c => !c.IsEmpty() && lookupPredicate.Invoke(new Expression(c.Value).CompareTo(lookup_value)))`

Does it really have to be this complicated? Does `.CellsUsed(XLCellsUsedOptions.Contents)` not return non-empty cells already? Can't we use `c => c.Value == lookup_value`? This way we'd lower the number of allocations (`new Expression(...)` for every analyzed cell).

Reply:
You're right about hte `IsEmpty()`, but the `Expression` way is necessary to correctly do comparison of different types and also case insensitive string comparison. Lessons I learnt from implementing `VLOOKUP`s. Maybe we should factor out `Expression.CompareTo()` in order instantiate fewer `Expression`s. I'd leave that for a separate PR though.